Move ICFY stat generation to Github Actions

blowery opened the pull request:
#### Changes proposed in this Pull Request

* Move ICFY stat generation to Github Actions

Requires:
- [x] https://github.com/Automattic/icfy/issues/23

blowery committed:
Move ICFY stat generation to Github Actions

blowery committed:
shell escaping is weird.

sirreal commented on .github/workflows/icfy-stats.yml:
```diff
@@ -0,0 +1,44 @@
+      with:
+        node-version: '>=12.15.0'
+    - name: Checkout code
+      uses: actions/checkout@master
```
That error looks like we'll need to do a deeper fetch here. Looks like it defaults to shallow:

https://github.com/actions/checkout/blob/f858c22e963bc60bc9d01c3d105c52a45a7deb6e/action.yml#L27-L29

sirreal commented on .github/workflows/icfy-stats.yml:
```diff
@@ -0,0 +1,44 @@
+      with:
+        node-version: '>=12.15.0'
+    - name: Checkout code
+      uses: actions/checkout@master
```
```suggestion
      with:
        fetch-depth: 0
```

sirreal replied:
This is probably a better recipe:

https://github.com/actions/checkout#fetch-all-tags

```yaml
- uses: actions/checkout@v2
- run: |
    git fetch --no-tags --prune --depth=1 origin +refs/heads/master:refs/remotes/origin/master
```

blowery committed:
fetch all the revs

Co-Authored-By: Jon Surrell <jon.surrell@automattic.com>

sirreal commented on .github/workflows/icfy-stats.yml:
```diff
@@ -0,0 +1,46 @@
+      with:
+        node-version: '>=12.15.0'
+    - name: Checkout code
+      uses: actions/checkout@master
```
Maybe this? We need to fetch the refs, not just the commits:
```suggestion
      uses: actions/checkout@v2
      run: git fetch --no-tags --prune --depth=1 origin +refs/heads/master:refs/remotes/origin/master
```
Maybe we can drop the with/fetch-depth as well.

https://github.com/actions/checkout#fetch-all-tags

blowery replied:
trying this one https://github.com/actions/checkout#fetch-all-history-for-all-tags-and-branches

blowery committed:
fetch all the remotes

blowery committed:
yaml is the best

sirreal committed:
Clean up whitespace

sirreal committed:
Try to get secret right

sirreal committed:
Turn off curl globbing

sirreal committed:
Split out env vars

sirreal committed:
Store icfy assets

sirreal committed:
Optimize head fetching

sirreal committed:
DROPME: Debug sha/branch

sirreal committed:
Add from:gh-action for ICFY

sirreal committed:
DROPME: Debug workers on GH actions

sirreal committed:
Limit workers to 2

Machines have 2 vCPUs:
https://docs.microsoft.com/en-us/azure/virtual-machines/dv2-dsv2-series#dsv2-series
`Standard_DS2_v2`

sirreal committed:
Files have moved, assume completion

sirreal committed:
Remove no fail comment

sirreal committed:
Unshallow

sirreal committed:
Allow error

sirreal committed:
Revert "DROPME: Debug workers on GH actions"

This reverts commit 07bcafa9ae11e770fc12d38ce6be0db3df2759e2.

sirreal committed:
fixup! Unshallow

sirreal committed:
Remove sha debug

sirreal committed:
Rename zip

sirreal commented:
This seems ready to me, ICFY will need some change to handle this.

blowery commented on .github/workflows/icfy-stats.yml:
```diff
@@ -0,0 +1,53 @@
+        ANCESTOR_SHA1=$(git merge-base HEAD origin/master)
+        CURRENT_BRANCH=$(git rev-parse --abbrev-ref HEAD)
+        curl -X POST --globoff                                                     \
+          "http://api.iscalypsofastyet.com:5000/submit-stats?secret=$ICFY_SECRET"  \
```
huh. it hated the `${{}}` syntax huh?

blowery replied:
hmmm, maybe the secret has/had bad URL characters in it?

sirreal replied:
We could inline the secret again without the env, the important bit is the `--globoff` option.

jsnajdr replied:
Yes, the secret contains characters like `%`, `|`, `}` or `*`. Definitely not URL and shell friendly 🙂

jsnajdr commented on .github/workflows/icfy-stats.yml:
```diff
@@ -0,0 +1,53 @@
+          -d '{
+                "payload": {
+                  "branch": "'"$CURRENT_BRANCH"'",
+                  "build_num": '"$GITHUB_RUN_NUMBER"',
```
I'm a bit confused by the build numbering. For the feb1484 commit, the `build_num` received by ICFY was `16`:
```
Received CircleCI success webhook notification: { payload:
   { branch: 'update/move-icfy-stats-to-github-action',
     build_num: 16,
     sha: 'feb1484add484d02d57e6728bd5ad0adf7529698',
     ancestor: '0c9f34fbee2020e8d771dce3c98a1439c30324e6',
     from: 'gh-action' } }
```

But the run numbers in GH web UI are 9 figures starting with 4 and the artifact download URL is:
```
https://github.com/Automattic/wp-calypso/suites/480255681/artifacts/2208657
```

sirreal replied:
Maybe this should be `GITHUB_RUN_ID`?

https://help.github.com/en/actions/configuring-and-managing-workflows/using-environment-variables

blowery replied:
Ugh. https://github.com/actions/upload-artifact/issues/50

Yeah, the run_id would let you query for artifacts: https://developer.github.com/v3/actions/artifacts/

jsnajdr replied:
Now after the latest commit the build number is `17`, the build number is `44892545` and the artifact URL is `https://github.com/Automattic/wp-calypso/suites/480418004/artifacts/2211405`. Not quite there yet, but much closer 🙂

jsnajdr replied:
```
curl https://api.github.com/repos/automattic/wp-calypso/actions/runs/44892545/artifacts
```
returns:
```
{
  "total_count": 1,
  "artifacts": [
    {
      "id": 2211405,
      "node_id": "MDg6QXJ0aWZhY3QyMjExNDA1",
      "name": "icfy",
      "size_in_bytes": 192025220,
      "url": "https://api.github.com/repos/Automattic/wp-calypso/actions/artifacts/2211405",
      "archive_download_url": "https://api.github.com/repos/Automattic/wp-calypso/actions/artifacts/2211405/zip",
      "expired": false,
      "created_at": "2020-02-25T13:04:29Z",
      "updated_at": "2020-02-25T13:04:30Z"
    }
  ]
}
```
I guess that with the build ID, getting to the artifact download URL became possible 🎉

jsnajdr commented on .github/workflows/icfy-stats.yml:
```diff
@@ -0,0 +1,53 @@
+    - uses: actions/upload-artifact@v1
+      with:
+        name: icfy
+        path: icfy-stats
```
Can we publish the artifacts as two plain files, `stats.json` and `chart.json`? That would be much more convenient for the ICFY client. It now downloads the files directly into memory. Now, it would need to unpack the ZIP to some tmp directory, read the JSON files from there, cleanup at the end...

sirreal replied:
This uses https://github.com/actions/upload-artifact. Docs make it sound like it will always be a zip:

> You must upload artifacts during a workflow run. GitHub provides two actions that you can use to upload and download build artifacts. Files uploaded to a workflow run are archived using the `.zip` format. For more information, see the actions/upload-artifact and download-artifact actions.

https://help.github.com/en/actions/configuring-and-managing-workflows/persisting-workflow-data-using-artifacts

blowery replied:
Doesn't seem like it. https://github.com/actions/upload-artifact

It wants a path to a directory and a name

blowery replied:
not sure it helps, but zips can be used as streams with https://www.npmjs.com/package/node-stream-zip

jsnajdr replied:
`node-stream-zip` can only read from files 🙁 I need something that can read from a generic stream. Will continue looking for another package.

jsnajdr replied:
Found https://www.npmjs.com/package/unzipper which works great so far :+1:

blowery committed:
add github_run_id

jsnajdr committed:
Pass from param in query, include only GITHUB_RUN_ID in payload

jsnajdr committed:
Change from=gh-action to from=github

jsnajdr reviewed (approved):
@sirreal @blowery @sgomes ICFY was updated and the latest build in this branch was performed by the GitHub Action and recorded correctly by ICFY.

The stats reported are wrong, because the base `master` commit doesn't have valid stats -- it failed with the well-known "out of memory" error.
